[1단계 - 로또 구현] 마갸(정찬영) 미션 제출합니다.

MyaGya opened the pull request:
안녕하세요 마갸입니다.
열심히 했는데 코드가 많이 부족한 것 같아요. 그래서 이번 미션까지는 양식대로 쓰지 않고 이대로 쓰려고 합니다. 어디에 시간을 투자했는지를 적으면 전부 다 5로 도배될 것 같아요, 기초가 많이 부족한 탓이죠

전체적으로 어려웠던 개념은 페어 프로그래밍에서 마크가 알려준 new를 쓰지 않는 생성 방식과 enum의 사용이었어요,
new를 쓰지 않고 유효값 체크를 해야 하는 객체에서 체크를 먼저 하게 만든 다음 생성을 하게 만들어서 메모리에 사용을 줄이는 방법이 이펙티브 자바 책에 적혀 있다고 하더라구요, 아직 이펙티브 자바를 볼 레벨은 아니라고 생각해서 틈틈이 얻는 개념만 공부하고 있습니다. 어쨋든 그런 구조를 기반으로 만들다 보니 뭐 하나 만드는게 쉽지 않아서 시간이 많이 늦어졌어요.

enum의 사용도 마찬가지였습니다. enum으로 연관되는 데이터를 묶는다는 개념은 알겠지만 코드를 진행하면서 특정 코드가 enum에서 담당해야하나 일반 객체에서 담당해야하나, 출력과 관련된 문구는 enum이 가지고 있어야 하는가 아닌가와 같이 헷갈리는 부분이 많아서 임의로 판단을 해서 구현을 했어요, 그 부분에 대해서 지적을 해주셨으면 고맙겠어요, 그 외에도 기초적인 실수가 엄청 많아서 리펙토링을 열심히 했는데도 아직 많이 코드가 허술해요, 지적 많이해주셨으면 감사하겠습니다.

binghe819 committed:
refactor: 사용하지 않는 파일 삭제

binghe819 committed:
docs: 기능 명세 작성

binghe819 committed:
feat: 로또 구입 금액 입력 기능

binghe819 committed:
feat: 로또 번호 생성 기능

binghe819 committed:
refactor: 로또 번호 생성 기능 리팩토링

binghe819 committed:
feat: 당첨 번호 입력 기능

MyaGya committed:
Merge branch 'step1' of https://github.com/binghe819/java-lotto into step1

binghe819 committed:
feat: MVC 패턴 구조 설계 및 로또 구입 금액 입력 기능 구현

MyaGya committed:
Merge branch 'step1' of https://github.com/binghe819/java-lotto into step1

MyaGya committed:
feat: 구매한 로또 번호들 출력 기능 추가

MyaGya committed:
feat: 당첨 번호 입력 기능 및 보너스 볼 입력 기능 추가

MyaGya committed:
feat: 보너스 번호 중복 체크 기능 추가

MyaGya committed:
feat: 로또 값 비교를 위한 테스트 코드 작성

MyaGya committed:
feat: 로또 순위 생성 기능 추가

MyaGya committed:
feat: 수익률 계산 기능 추가

MyaGya committed:
feat: 로또 결과 출력 구문 생성

MyaGya committed:
feat: 로또 수익률 출력 구현

MyaGya committed:
feat: 에러 메세지 출력 테스트

MyaGya committed:
feat: 에러 메세지를 출력하고 프로그램을 재시작하도록 구현

MyaGya committed:
refactor: 전체 컨트롤러 메소드 분할

MyaGya committed:
refactor: 로또 번호를 상수값으로 변경하고 한곳에서만 참조하도록 변경

MyaGya committed:
refactor: 로또 최대 갯수를 6으로 변경

LottoNumber 클래스에서만 참조하도록 변경

MyaGya committed:
refactor: 로또 갯수 관련 에러 메세지를 상수값과 연동하도록 변경

MyaGya committed:
refactor: 메세지 하드코딩 상수화

MyaGya committed:
refactor: 로또 값을 한 클래스에서 전담하도록 변경

MyaGya committed:
docs: README 문서 메인으로 이동

MyaGya committed:
refactor: splitter 상수로 변경

seok-2-o commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+        List<LottoNumber> lottoNumbers = numbers.stream()
+            .map(number -> LottoNumber.valueOf(number))
+            .collect(Collectors.toList());
+        return Lotto.generatedBy(() -> lottoNumbers);
```
```suggestion
        return new Lotto(lottoNumbers)
```

seok-2-o commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+
+public class Lotto {
+
+    public static final String NUM_LOTTO_ERROR_MESSAGE = "로또 번호는 %d개의 숫자로 이루어져야 합니다.";
```
public 으로 선언하신 이유가 있을까요?

MyaGya replied:
객체들을 자동으로 올리도록 intellij 명령어를 사용했었는데 당연히 private로 올라가는 줄 알고 자세히 확인하지 않았습니다. 주의할게요

seok-2-o commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+
+    public static Lotto generatedBy(LottoGenerator lottoGenerator) {
+        List<LottoNumber> lottoNumbers = lottoGenerator.generateLottoNumbers();
+        validateLotto(lottoNumbers);
```
생성자에서 벨리데이션을 하는것과 정적팩토리에서 하는것 어디서 벨리데이션을 수행하는 것이 좋을까요? 🤔

MyaGya replied:
이부분은 처음에 남겨주신 로그와 덩달아서 헷갈리는 부분이라서 조금 가르쳐주셨으면 감사하겠습니다.
`Lotto `클래스에 `generatedBy`는 오버로딩으로 두 개의 메소드가 존재하도록 구현하였습니다. 그리고 최종적으로 코드의 `generatedBy` 의 형태로 바뀌어서 요청이 들어옵니다. 그래서 저는 정적팩토리라는 개념이 JAVA에서 new로 바로 데이터를 만들어버리면 객체 검증 단계에서 문제가 발생하기 때문에 new의 부분이 아닌 팩토리 단계에서 미리 값을 검증해야한다고 생각했어요. 즉,

1. 각 메소드마다 예외처리를 따로 해주고 new로 넘기면 메모리의 낭비가 생긴다, 따라서 모든 생성 방식을 가장 하단에 있는 `generatedBy` 로 데이터를 가공해서 보내고 해당 메소드에서 유효값을 검증하기만 하고 생성자는 말 그대로 생성하는 기능만 담당해야할 것이다.
라고 판단을 했습니다. 이 부분이 이상할까요?

seok-2-o commented on src/main/java/lotto/domain/lotto/LottoRank.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Arrays;
+
+public enum LottoRank {
+    NONE(0, 0),
+    SEVENTH(1, 0),
+    SIXTH(2, 0),
```
5등 이하를 분리하는것은 의미가 없을 것 같아요.

MyaGya replied:
이 부분은 억지로 분리하지 않아도 되지만 가치가 0이 되는 부분을 한꺼번에 묶어서 처리하도록 해서 put 부분에서 편의를 위해서 이렇게 만들었어요, 크게 문제가 될까요?
로또 결과를 출력하는 `LottoResult` 에서
`rank.put(lottoRank, rank.get(lottoRank) + 1);` 이 부분입니다. 이 로직을 처리하려면 그냥 경우의 수가 다 있는편이 낫다고 생각했어요.

seok-2-o replied:
불필요한 상태는 나중에 혼란을 가중시킨다고 생각해요. 예를들어 MISS (꽝) 의 갯수를 출력하는 로직을 만들어야한다면
추가작업이 필요할 것 같아요.

seok-2-o commented on src/main/java/lotto/domain/lotto/LottoResult.java:
```diff
@@ -0,0 +1,52 @@
+
+public class LottoResult {
+
+    public static final int RATE = 100;
```
public 으로 선언하신 이유가 있나요 ?

MyaGya replied:
위와 동일하게 기능을 잘못 알고 있었어요, 주의할게요

seok-2-o commented on src/main/java/lotto/domain/lotto/LottoResult.java:
```diff
@@ -0,0 +1,52 @@
+        return this.rank;
+    }
+
+    public void add(Lotto lotto, WinningLotto winningLotto) {
```
메소드 명으로 해당 행위를 설명하기에는 다소 부족한 것 같아요.

MyaGya replied:
로또를 체크해서 몇 개나 일치하는지를 구하는 기능에 맞게 `checkWinningLotto` 로 변경하였습니다.

seok-2-o commented on src/main/java/lotto/domain/lotto/LottoResult.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void add(Lotto lotto, WinningLotto winningLotto) {
+        int count = (int) lotto.getNumbers().stream()
+            .filter(number -> winningLotto.getLotto().getNumbers().contains(number)).count();
```
값을 외부로 꺼내서 비교하고 있네요. 이는 올바른 객체지향방법이 아닌것 같아요
`LottoRank  rank = lotto.check(winningLotto)`;
와 같은 꼴로 리펙토링 진행해주세요.

MyaGya replied:
이부분은 고민을 되게 많이했습니다. 시간에 쫒겨 제출해서 제대로 책임 분산을 시켜주지 못했엇습니다.
`Lotto` 클래스에서 로또 일치, 보너스 일치를 담당하고
`LottoResult` 클래스에서 일치된 로또 삽입 연산을 수행해주도록 분산시켰습니다.

seok-2-o commented on src/main/java/lotto/domain/lotto/Lottos.java:
```diff
@@ -0,0 +1,20 @@
+
+    public Integer getNumLotto(){
+        return values.size();
+    }
```
```suggestion
    public Integer size(){
        return values.size();
    }
```

MyaGya replied:
이 부분은 `size`로 할까 고민을 하다가 `getNumLotto`라는 이름이 더 직관적이라고 생각해서 `getNumLotto`라고 사용해주었습니다. `Lottos`가 일급 컬렉션으로 만들면서 사용하는 입장에서 `Lottos`의 `size` 보다 `Lottos`의 `getNumLotto` 가 더 쓰기 좋을거라 생각했습니다. 우선 size로 바꿧습니다.

seok-2-o commented on src/main/java/lotto/domain/lotto/Lottos.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public class Lottos {
```
해당 객체가 하는 역할이 무엇인가요 ?

MyaGya replied:
사용자가 구매한 로또의 묶음을 lottos로 따로 분리해서 사용해주는게 편할거라 생각해서 분리했습니다. 이상한 부분이 있나요?

seok-2-o replied:
묶어주는 역할외에 아무른 역할이 없는것 같아요 
getLottos 을 통해서 바로 상태를 노출하고 있네요
```java
LottoController.java
       for (Lotto lotto : purchasedLottos.getLottos()) {
            lottoResult.checkWinningLotto(lotto, winningLotto);
        }
```
해당 로직이 getter 로 상태를 노출하지 않고 와 같은 꼴로 변경되면 해당 객체도 책임을 가질 수 있을 것 같아요.

purchasedLottos.check(winningLotto);

seok-2-o commented on src/main/java/lotto/domain/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public static WinningLotto generatedBy(Lotto lotto, LottoNumber bonus) {
+        validateWinningLotto(lotto, bonus);
```
```suggestion
        validateBonus(lotto, bonus);
```
와는 무슨 차이가 있나요 ?

MyaGya replied:
당장 코드상에 의미는 없지만 구현 도중에 생각해보니 `WinningLotto` 를 만들기 위한 예외처리 단계에서 `validateWinningLotto` 로 모든 예외처리를 관리하는 편이 더 좋지 않을까 생각해서 시험적으로 만들어보았습니다. 결과적으로 예외처리 해야 할 내용이 하나 밖에 없어서 코드를 깔끔하게 정리했어야 하는데 놓쳐버렸네요 조심하겠습니다.

seok-2-o reviewed (changes requested):
안녕하세요 마갸. 이번 미션을 함께하게 된 미립입니다. 반갑습니다 🙇 

미션 하느라 고생 많으셨어요. 
객체지향에서는 객체의 값을 꺼내 어떤 행위를 통해 문제를 해결하기 보다 객체간의 관계를 통해 문제를 해결하도록 해야하는데
이부분이 반영되지 않은 것 같아요. 또한 일급컬렉션은 잘 만들어주셨지만 일급컬렉션이 하는 일이 없네요. 

객체들이 자기가 할 일을 할 수 있도록 개선해보면 좋겠네요. 

이해하기 어렵거나 막히는 부분이 있다면 편하게 메세지 주세요. 같이 고민해보아요 🔥

MyaGya committed:
refactor: 접근제어자를 알맞게 변경

before:
public 이지만 클래스 내부에서만 사용함

after:
private로 변경하여 클래스 내부에서만 계속 사용하도록 명시

MyaGya committed:
refactor: 접근제어자를 알맞게 변경

before:
public 이지만 클래스 내부에서만 사용함

after:
private로 변경하여 클래스 내부에서만 계속 사용하도록 명시

MyaGya committed:
refactor: 메소드 명을 기능에 더 맞게 변경

before:
메소드 명이 add 이고 로또를 계산해서 더하는 기능에는 맞지 않음

after:
checkWinningLotto 로 변경

MyaGya committed:
refactor: 로또 결과 출력 기능 책임 변경

before:
LottoResult 에서 로또 일치 수, 보너스 일치 수, 삽입 연산까지 모두 담당

after:
Lotto 에서 로또 일치수, 보너스 일치수 확인 후 삽입 연산만 LottoResult 에서 담당

MyaGya committed:
refactor: 로또 일치 갯수 테스트 코드 수정

before:
일치 수를 찾는 로직을 구현 후 메소드로 만들었으나 로직만 테스트 코드에 남아있음

after:
로직을 lottoResult 에서 가져와서 사용

MyaGya committed:
refactor: Lottos 일급 컬렉션의 메소드 명 변경

before:
getNumLotto

after:
size

MyaGya committed:
refactor: 의미없는 메소드 제거

MyaGya commented:
많은 피드백 감사합니다. 이번 과제는 특히 시간 제한이 있어 더 어려웠던 것 같아요. 지적한 부분을 보니 분명히 사전에는 이 부분을 고려했는데 짤 때는 시간에 쫒겨 까먹어 버린다던지, 기능 구현에 너무 집중해서 여러 책임을 함께 떠맡는다던지, 다양한 문제가 생겼네요. 사실 남들보다는 늦지만 이제야 조금 TDD에 개념이 잡히는 것 같아요, 여러 테스트를 만들다 보니 테스트만을 위한 코드가 생기는데 좀 더 공부해야 테스트만을 위한 코드가 아니라 테스트를 하기 쉬운 코드를 만들 수 있을 것 같아요. 객체지향에 대한 개념은 아직 아리송해서 지적을 받아야지 눈에 보이네요. 더 시간을 투자해야 할 것 같아요.

seok-2-o reviewed (approved):
피드백 반영잘해주셨네요 :) 
지금도 잘하고 계시는것 같으니 너무 조급해하지 않으셔도 될 것 같습니다.
코멘트에 추가로 몇가지 코멘트 남겨놓았으니 다음 미션 진행하면서 고민해보면 좋을것 같아요.

이어서 다음 미션 진행해주세요.
